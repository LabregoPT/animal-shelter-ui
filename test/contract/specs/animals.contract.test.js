import { provider } from "../config/init-pact";
import { Matchers } from "@pact-foundation/pact";
import { AnimalController } from "../../../controllers";

describe('Animal Service', () => {
	beforeAll(async () => {
		await provider.setup();
	});

	describe("When a request to list all animals is made", () => {
		beforeEach(async () => {
			await provider.addInteraction({
				uponReceiving: "a request to list all animals",
				state: "has animals",
				withRequest: {
					method: "GET",
					path: "/animals",
				},
				willRespondWith: {
					status: 200,
					body: Matchers.eachLike({
						name: Matchers.like("manchas"),
						breed: Matchers.like("Bengali"),
						gender: Matchers.like("Female"),
						vaccinated: Matchers.boolean(true),
					}),
				},
			});
		});
		test("should return the correct data", async () => {
			const response = await AnimalController.list();
			expect(response.data).toMatchSnapshot();
			await provider.verify();
		});
	});

	describe("When a request to get an animal is made", () => {
		beforeEach(async () => {
			await provider.addInteraction({
				uponReceiving: "a request to get an animal",
				state: "has an animal",
				withRequest: {
					method: "GET",
					path: "/animals/Kiara",
				},
				willRespondWith: {
					status: 200,
					body: {
						name: Matchers.like("Kiara"),
						breed: Matchers.like("Bengali"),
						gender: Matchers.like("Female"),
						vaccinated: Matchers.boolean(true),
					},
				},
			});
		});
		test("should return the correct data", async () => {
			const response = await AnimalController.getAnimal("Kiara");
			expect(response.data).toMatchSnapshot();
			await provider.verify();
		});
	});

	describe("When a request to add an animal is made", () => {
		let testBody = {
			name: "Kiara",
			breed: "Criolla",
			gender: "Female",
			vaccinated: false,
		}
		beforeEach(async () => {
			await provider.addInteraction({
				uponReceiving: "a request to add an animal",
				state: "has an animal to update",
				withRequest: {
					method: "POST",
					path: "/animals",
					body: testBody
				},
				willRespondWith: {
					status: 201,
					body: testBody
				},
			});
		});
		test("should return the correct data", async () => {
			const response = await AnimalController.register(testBody);
			expect(response.data).toMatchSnapshot();
			await provider.verify();
		});
	});

	describe("When a request to update an animal is made", () => {
		let testBody = {
			name: "Kiara",
			breed: "Criolla",
			gender: "Female",
			vaccinated: true,
			vaccines: ["Polio","Rabia"]
		}
		beforeEach(async () => {
			await provider.addInteraction({
				uponReceiving: "a request to update an animal",
				state: "has an animal to update",
				withRequest: {
					method: "PUT",
					path: "/animals/Kiara",
					body: testBody
				},
				willRespondWith: {
					status: 200,
					body: {
						id: Matchers.like(6),
						name: Matchers.like("Kiara"),
						breed: Matchers.like("Criolla"),
						gender: Matchers.like("Female"),
						vaccinated: Matchers.boolean(true),
						vaccines: Matchers.like(["Polio", "Rabia"])
					}
				},
			});
		});
		test("should return the correct data", async () => {
			const response = await AnimalController.updateAnimal("Kiara",testBody);
			expect(response.data).toMatchSnapshot();
			await provider.verify();
		});
	});

	
	afterAll(() => provider.finalize());
});