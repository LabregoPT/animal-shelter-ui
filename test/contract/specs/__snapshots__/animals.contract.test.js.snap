// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animal Service When a request to add an animal is made should return the correct data 1`] = `
Object {
  "breed": "Criolla",
  "gender": "Female",
  "name": "Kiara",
  "vaccinated": false,
}
`;

exports[`Animal Service When a request to delete an animal is made should return the correct data 1`] = `""`;

exports[`Animal Service When a request to get an animal is made should return the correct data 1`] = `
Object {
  "breed": "Bengali",
  "gender": "Female",
  "name": "Kiara",
  "vaccinated": true,
}
`;

exports[`Animal Service When a request to list all animals is made should return the correct data 1`] = `
Array [
  Object {
    "breed": "Bengali",
    "gender": "Female",
    "name": "manchas",
    "vaccinated": true,
  },
]
`;

exports[`Animal Service When a request to update an animal is made should return the correct data 1`] = `
Object {
  "breed": "Criolla",
  "gender": "Female",
  "id": 6,
  "name": "Kiara",
  "vaccinated": true,
  "vaccines": Array [
    "Polio",
    "Rabia",
  ],
}
`;
